# Vehicle loan default prediction

# Introduction

Financial institutions incur significant losses due to the default of vehicle loans. This has led to the tightening up of vehicle loan underwriting and increased vehicle loan rejection rates. The need for a better credit risk scoring model is also raised by these institutions. This warrants a study to estimate the determinants of vehicle loan default. A financial institution has hired you to accurately predict the probability of loanee/borrower defaulting on a vehicle loan in the first EMI (Equated Monthly Instalments) on the due date.

# Data source

The data is obtained from third party data website Kaggle. The data comes from L&T finance services.

Following Information regarding the loan and loanee are provided in the datasets:
* Loanee Information (Demographic data like age, Identity proof etc.)
* Loan Information (Disbursal details, loan to value ratio etc.)
* Bureau data & history (Bureau score, number of active accounts, the status of other loans, credit history etc.)

Link of the data source - [L&T vehicle loan default data](https://www.kaggle.com/datasets/mamtadhaker/lt-vehicle-loan-default-prediction)

# Data import

In [141]:
import pandas as pd
import re

loan_train_df = pd.read_csv("./data/train.csv")
loan_test_df = pd.read_csv("./data/test.csv")

pd.set_option('display.max_columns', None)

# Structure of data

In [142]:
loan_train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 233154 entries, 0 to 233153
Data columns (total 41 columns):
 #   Column                               Non-Null Count   Dtype  
---  ------                               --------------   -----  
 0   UniqueID                             233154 non-null  int64  
 1   disbursed_amount                     233154 non-null  int64  
 2   asset_cost                           233154 non-null  int64  
 3   ltv                                  233154 non-null  float64
 4   branch_id                            233154 non-null  int64  
 5   supplier_id                          233154 non-null  int64  
 6   manufacturer_id                      233154 non-null  int64  
 7   Current_pincode_ID                   233154 non-null  int64  
 8   Date.of.Birth                        233154 non-null  object 
 9   Employment.Type                      225493 non-null  object 
 10  DisbursalDate                        233154 non-null  object 
 11  State_ID     

There are total 41 columns and 233154 rows.

In [143]:
loan_train_df.head(7)

UniqueID  disbursed_amount  asset_cost    ltv  branch_id  supplier_id   
0    420825             50578       58400  89.55         67        22807  \
1    537409             47145       65550  73.23         67        22807   
2    417566             53278       61360  89.63         67        22807   
3    624493             57513       66113  88.48         67        22807   
4    539055             52378       60300  88.39         67        22807   
5    518279             54513       61900  89.66         67        22807   
6    529269             46349       61500  76.42         67        22807   

   manufacturer_id  Current_pincode_ID Date.of.Birth Employment.Type   
0               45                1441      01-01-84        Salaried  \
1               45                1502      31-07-85   Self employed   
2               45                1497      24-08-85   Self employed   
3               45                1501      30-12-93   Self employed   
4               45                1495      09-12-77   Self employed   
5               45                1501      08-09-90   Self employed   
6               45                1502      01-06-88        Salaried   

  DisbursalDate  State_ID  Employee_code_ID  MobileNo_Avl_Flag  Aadhar_flag   
0      03-08-18         6              1998                  1            1  \
1      26-09-18         6              1998                  1            1   
2      01-08-18         6              1998                  1            1   
3      26-10-18         6              1998                  1            1   
4      26-09-18         6              1998                  1            1   
5      19-09-18         6              1998                  1            1   
6      23-09-18         6              1998                  1            1   

   PAN_flag  VoterID_flag  Driving_flag  Passport_flag  PERFORM_CNS.SCORE   
0         0             0             0              0                  0  \
1         0             0             0              0                598   
2         0             0             0              0                  0   
3         0             0             0              0                305   
4         0             0             0              0                  0   
5         0             0             0              0                825   
6         0             0             0              0                  0   

  PERFORM_CNS.SCORE.DESCRIPTION  PRI.NO.OF.ACCTS  PRI.ACTIVE.ACCTS   
0   No Bureau History Available                0                 0  \
1                 I-Medium Risk                1                 1   
2   No Bureau History Available                0                 0   
3              L-Very High Risk                3                 0   
4   No Bureau History Available                0                 0   
5               A-Very Low Risk                2                 0   
6   No Bureau History Available                0                 0   

   PRI.OVERDUE.ACCTS  PRI.CURRENT.BALANCE  PRI.SANCTIONED.AMOUNT   
0                  0                    0                      0  \
1                  1                27600                  50200   
2                  0                    0                      0   
3                  0                    0                      0   
4                  0                    0                      0   
5                  0                    0                      0   
6                  0                    0                      0   

   PRI.DISBURSED.AMOUNT  SEC.NO.OF.ACCTS  SEC.ACTIVE.ACCTS  SEC.OVERDUE.ACCTS   
0                     0                0                 0                  0  \
1                 50200                0                 0                  0   
2                     0                0                 0                  0   
3                     0                0                 0                  0   
4                     0                0                

Checking the test dataset.

In [144]:
loan_test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112392 entries, 0 to 112391
Data columns (total 40 columns):
 #   Column                               Non-Null Count   Dtype  
---  ------                               --------------   -----  
 0   UniqueID                             112392 non-null  int64  
 1   disbursed_amount                     112392 non-null  int64  
 2   asset_cost                           112392 non-null  int64  
 3   ltv                                  112392 non-null  float64
 4   branch_id                            112392 non-null  int64  
 5   supplier_id                          112392 non-null  int64  
 6   manufacturer_id                      112392 non-null  int64  
 7   Current_pincode_ID                   112392 non-null  int64  
 8   Date.of.Birth                        112392 non-null  object 
 9   Employment.Type                      108949 non-null  object 
 10  DisbursalDate                        112392 non-null  object 
 11  State_ID     

Test set contains 40 columns as it doesn't `loan_default` column.

# Preprocessing

In [145]:
def fill_emp_nan(data_df):
    data_df = data_df.copy()
    
    # replace "nan" with "Unemployed"
    data_df["Employment.Type"] = data_df["Employment.Type"].fillna("Unemployed")
    return data_df


In [146]:
def drop_cols(data_df, columns):
    data_df = data_df.copy()
    
    # droping UniqueID, supplier_id, Current_pincode_ID, Employee_code_ID
    data_df = data_df.drop(columns, axis=1)
    
    return data_df

In [147]:
def extract_month_year(data_df, column):
    data_df = data_df.copy()
    pattern_year = "([\d]+)yrs"
    pattern_month = "([\d]+)mon"
    data_df[column + "_year"] = data_df[column].apply(lambda x: re.findall(pattern_year, x)[0])
    data_df[column + "_month"] = data_df[column].apply(lambda x: re.findall(pattern_month, x)[0])
    data_df[column + "_year"] = data_df[column + "_year"].astype("int")
    data_df[column + "_month"] = data_df[column + "_month"].astype("int")
    data_df = data_df.drop(column, axis=1)
    return data_df

In [148]:
def one_hot_encode(data_df, columns):
    for column in columns:
        dummies_df = pd.get_dummies(data_df[column], prefix=column, dtype="int")
        data_df = pd.concat([data_df, dummies_df], axis=1)
        data_df = data_df.drop(column, axis=1)
    return data_df

Creating a separate label set "Y_train".

In [149]:
Y_train = loan_train_df["loan_default"]

In [150]:
Y_train

0         0
1         1
2         0
3         1
4         1
         ..
233149    0
233150    0
233151    0
233152    0
233153    0
Name: loan_default, Length: 233154, dtype: int64

In [151]:
# creating a separate set for training example
X_train = loan_train_df.copy()
X_train = X_train.drop("loan_default", axis=1)

Checking for the missing values.

In [152]:
X_train.isna().sum()

UniqueID                                  0
disbursed_amount                          0
asset_cost                                0
ltv                                       0
branch_id                                 0
supplier_id                               0
manufacturer_id                           0
Current_pincode_ID                        0
Date.of.Birth                             0
Employment.Type                        7661
DisbursalDate                             0
State_ID                                  0
Employee_code_ID                          0
MobileNo_Avl_Flag                         0
Aadhar_flag                               0
PAN_flag                                  0
VoterID_flag                              0
Driving_flag                              0
Passport_flag                             0
PERFORM_CNS.SCORE                         0
PERFORM_CNS.SCORE.DESCRIPTION             0
PRI.NO.OF.ACCTS                           0
PRI.ACTIVE.ACCTS                

Only `Employment.Type` has missing values and it contains 7661 missing values.

Let's check for the type of values `Employment.Type` consists of.

In [153]:
loan_test_df.isna().sum()

UniqueID                                  0
disbursed_amount                          0
asset_cost                                0
ltv                                       0
branch_id                                 0
supplier_id                               0
manufacturer_id                           0
Current_pincode_ID                        0
Date.of.Birth                             0
Employment.Type                        3443
DisbursalDate                             0
State_ID                                  0
Employee_code_ID                          0
MobileNo_Avl_Flag                         0
Aadhar_flag                               0
PAN_flag                                  0
VoterID_flag                              0
Driving_flag                              0
Passport_flag                             0
PERFORM_CNS.SCORE                         0
PERFORM_CNS.SCORE.DESCRIPTION             0
PRI.NO.OF.ACCTS                           0
PRI.ACTIVE.ACCTS                

Test set aslo contains missing values in `Employment.Type` column.

In [154]:
loan_train_df["Employment.Type"].unique()

array(['Salaried', 'Self employed', nan], dtype=object)

So there are three categories "Salaried", "Self employed", "nan". As only `Employment.Type` has missing values and it has two categories(Salaried, Self employed) for employed people, it is safe to assume that "nan" refers to unemployed people.

In [155]:
X_train = fill_emp_nan(X_train)
X_test = fill_emp_nan(loan_test_df)

In [156]:
print("missing values in training data: ", X_train.isna().sum().sum())
print("missing values in test data: ", X_test.isna().sum().sum())

missing values in training data:  0
missing values in test data:  0


In [157]:
X_train.head()

UniqueID  disbursed_amount  asset_cost    ltv  branch_id  supplier_id   
0    420825             50578       58400  89.55         67        22807  \
1    537409             47145       65550  73.23         67        22807   
2    417566             53278       61360  89.63         67        22807   
3    624493             57513       66113  88.48         67        22807   
4    539055             52378       60300  88.39         67        22807   

   manufacturer_id  Current_pincode_ID Date.of.Birth Employment.Type   
0               45                1441      01-01-84        Salaried  \
1               45                1502      31-07-85   Self employed   
2               45                1497      24-08-85   Self employed   
3               45                1501      30-12-93   Self employed   
4               45                1495      09-12-77   Self employed   

  DisbursalDate  State_ID  Employee_code_ID  MobileNo_Avl_Flag  Aadhar_flag   
0      03-08-18         6              1998                  1            1  \
1      26-09-18         6              1998                  1            1   
2      01-08-18         6              1998                  1            1   
3      26-10-18         6              1998                  1            1   
4      26-09-18         6              1998                  1            1   

   PAN_flag  VoterID_flag  Driving_flag  Passport_flag  PERFORM_CNS.SCORE   
0         0             0             0              0                  0  \
1         0             0             0              0                598   
2         0             0             0              0                  0   
3         0             0             0              0                305   
4         0             0             0              0                  0   

  PERFORM_CNS.SCORE.DESCRIPTION  PRI.NO.OF.ACCTS  PRI.ACTIVE.ACCTS   
0   No Bureau History Available                0                 0  \
1                 I-Medium Risk                1                 1   
2   No Bureau History Available                0                 0   
3              L-Very High Risk                3                 0   
4   No Bureau History Available                0                 0   

   PRI.OVERDUE.ACCTS  PRI.CURRENT.BALANCE  PRI.SANCTIONED.AMOUNT   
0                  0                    0                      0  \
1                  1                27600                  50200   
2                  0                    0                      0   
3                  0                    0                      0   
4                  0                    0                      0   

   PRI.DISBURSED.AMOUNT  SEC.NO.OF.ACCTS  SEC.ACTIVE.ACCTS  SEC.OVERDUE.ACCTS   
0                     0                0                 0                  0  \
1                 50200                0                 0                  0   
2                     0                0                 0                  0   
3                     0                0                 0                  0   
4                     0                0                 0                  0   

   SEC.CURRENT.BALANCE  SEC.SANCTIONED.AMOUNT  SEC.DISBURSED.AMOUNT   
0                    0                      0                     0  \
1                    0                      0                     0   
2                    0                      0                     0   
3                    0                      0                     0   
4                    0                      0                     0   

   PRIMARY.INSTAL.AMT  SEC.INSTAL.AMT  NEW.ACCTS.IN.LAST.SIX.MONTHS   
0                   0               0                             0  \
1                1991               0                             0   
2                   0               0                             0   
3                  31               0                             0   
4                   0               0                             0   

   D

Let's first focus on categorical columns(mainly columns with many categorical values) leaving out the binary categorical columns.

In [158]:
cat_columns = ["UniqueID", "branch_id", "supplier_id", "manufacturer_id", "Current_pincode_ID", "Employment.Type", "State_ID", "Employee_code_ID", "PERFORM_CNS.SCORE.DESCRIPTION"]

{column: len(X_train[column].unique()) for column in cat_columns}

{'UniqueID': 233154,
 'branch_id': 82,
 'supplier_id': 2953,
 'manufacturer_id': 11,
 'Current_pincode_ID': 6698,
 'Employment.Type': 3,
 'State_ID': 22,
 'Employee_code_ID': 3270,
 'PERFORM_CNS.SCORE.DESCRIPTION': 20}

Drop `UniqueID` column.

Drop `supplier_id`, `Current_pincode_ID`, `Employee_code_ID` because of high cardinality. High cardinality columns will make training the model take too much time. For this project, it's better to avoid that situation.

In [159]:
X_train = drop_cols(X_train, ["UniqueID", "supplier_id", "Current_pincode_ID", "Employee_code_ID"])
X_test = drop_cols(X_test, ["UniqueID", "supplier_id", "Current_pincode_ID", "Employee_code_ID"])

In [160]:
X_train.head()

disbursed_amount  asset_cost    ltv  branch_id  manufacturer_id   
0             50578       58400  89.55         67               45  \
1             47145       65550  73.23         67               45   
2             53278       61360  89.63         67               45   
3             57513       66113  88.48         67               45   
4             52378       60300  88.39         67               45   

  Date.of.Birth Employment.Type DisbursalDate  State_ID  MobileNo_Avl_Flag   
0      01-01-84        Salaried      03-08-18         6                  1  \
1      31-07-85   Self employed      26-09-18         6                  1   
2      24-08-85   Self employed      01-08-18         6                  1   
3      30-12-93   Self employed      26-10-18         6                  1   
4      09-12-77   Self employed      26-09-18         6                  1   

   Aadhar_flag  PAN_flag  VoterID_flag  Driving_flag  Passport_flag   
0            1         0             0             0              0  \
1            1         0             0             0              0   
2            1         0             0             0              0   
3            1         0             0             0              0   
4            1         0             0             0              0   

   PERFORM_CNS.SCORE PERFORM_CNS.SCORE.DESCRIPTION  PRI.NO.OF.ACCTS   
0                  0   No Bureau History Available                0  \
1                598                 I-Medium Risk                1   
2                  0   No Bureau History Available                0   
3                305              L-Very High Risk                3   
4                  0   No Bureau History Available                0   

   PRI.ACTIVE.ACCTS  PRI.OVERDUE.ACCTS  PRI.CURRENT.BALANCE   
0                 0                  0                    0  \
1                 1                  1                27600   
2                 0                  0                    0   
3                 0                  0                    0   
4                 0                  0                    0   

   PRI.SANCTIONED.AMOUNT  PRI.DISBURSED.AMOUNT  SEC.NO.OF.ACCTS   
0                      0                     0                0  \
1                  50200                 50200                0   
2                      0                     0                0   
3                      0                     0                0   
4                      0                     0                0   

   SEC.ACTIVE.ACCTS  SEC.OVERDUE.ACCTS  SEC.CURRENT.BALANCE   
0                 0                  0                    0  \
1                 0                  0                    0   
2                 0                  0                    0   
3                 0                  0                    0   
4                 0                  0                    0   

   SEC.SANCTIONED.AMOUNT  SEC.DISBURSED.AMOUNT  PRIMARY.INSTAL.AMT   
0                      0                     0                   0  \
1                      0                     0                1991   
2                      0                     0                   0   
3                      0                     0                  31   
4                      0                     0                   0   

   SEC.INSTAL.AMT  NEW.ACCTS.IN.LAST.SIX.MONTHS   
0               0                             0  \
1               0                             0   
2               0                             0   
3               0                             0   
4               0                             0   

   DELINQUENT.ACCTS.IN.LAST.SIX.MONTHS AVERAGE.ACCT.AGE CREDIT.HISTORY.LENGTH   
0                                    0        0yrs 0mon             0yrs 0mon  \
1                                    1       1yrs 11mon            1yrs 11mon   
2                                    0        0yrs 0mon             0yrs 0mon   
3                                    0        

`Date.of.Birth`, `DisbursalDate` can also be dropped as these fields cannot be used for generalization.

In [161]:
# remove "Date.of.Birth", "DisbursalDate" columns
X_train = drop_cols(X_train, ["Date.of.Birth", "DisbursalDate"])
X_test = drop_cols(X_test, ["Date.of.Birth", "DisbursalDate"])
X_train.head()

disbursed_amount  asset_cost    ltv  branch_id  manufacturer_id   
0             50578       58400  89.55         67               45  \
1             47145       65550  73.23         67               45   
2             53278       61360  89.63         67               45   
3             57513       66113  88.48         67               45   
4             52378       60300  88.39         67               45   

  Employment.Type  State_ID  MobileNo_Avl_Flag  Aadhar_flag  PAN_flag   
0        Salaried         6                  1            1         0  \
1   Self employed         6                  1            1         0   
2   Self employed         6                  1            1         0   
3   Self employed         6                  1            1         0   
4   Self employed         6                  1            1         0   

   VoterID_flag  Driving_flag  Passport_flag  PERFORM_CNS.SCORE   
0             0             0              0                  0  \
1             0             0              0                598   
2             0             0              0                  0   
3             0             0              0                305   
4             0             0              0                  0   

  PERFORM_CNS.SCORE.DESCRIPTION  PRI.NO.OF.ACCTS  PRI.ACTIVE.ACCTS   
0   No Bureau History Available                0                 0  \
1                 I-Medium Risk                1                 1   
2   No Bureau History Available                0                 0   
3              L-Very High Risk                3                 0   
4   No Bureau History Available                0                 0   

   PRI.OVERDUE.ACCTS  PRI.CURRENT.BALANCE  PRI.SANCTIONED.AMOUNT   
0                  0                    0                      0  \
1                  1                27600                  50200   
2                  0                    0                      0   
3                  0                    0                      0   
4                  0                    0                      0   

   PRI.DISBURSED.AMOUNT  SEC.NO.OF.ACCTS  SEC.ACTIVE.ACCTS  SEC.OVERDUE.ACCTS   
0                     0                0                 0                  0  \
1                 50200                0                 0                  0   
2                     0                0                 0                  0   
3                     0                0                 0                  0   
4                     0                0                 0                  0   

   SEC.CURRENT.BALANCE  SEC.SANCTIONED.AMOUNT  SEC.DISBURSED.AMOUNT   
0                    0                      0                     0  \
1                    0                      0                     0   
2                    0                      0                     0   
3                    0                      0                     0   
4                    0                      0                     0   

   PRIMARY.INSTAL.AMT  SEC.INSTAL.AMT  NEW.ACCTS.IN.LAST.SIX.MONTHS   
0                   0               0                             0  \
1                1991               0                             0   
2                   0               0                             0   
3                  31               0                             0   
4                   0               0                             0   

   DELINQUENT.ACCTS.IN.LAST.SIX.MONTHS AVERAGE.ACCT.AGE CREDIT.HISTORY.LENGTH   
0                                    0        0yrs 0mon             0yrs 0mon  \
1                                    1       1yrs 11mon            1yrs 11mon   
2                                    0        0yrs 0mon             0yrs 0mon   
3                                    0        0yrs 8mon             1yrs 3mon   
4                                    0        0yrs 0mon             0yrs 0mon   

   NO.OF_INQUIRIES  
0                0  
1                0  
2                0 

In [162]:
# extract year and month in separate columns from "AVERAGE.ACCT.AGE"
X_train = extract_month_year(X_train, "AVERAGE.ACCT.AGE")
X_test = extract_month_year(X_test, "AVERAGE.ACCT.AGE")
X_train.head()

disbursed_amount  asset_cost    ltv  branch_id  manufacturer_id   
0             50578       58400  89.55         67               45  \
1             47145       65550  73.23         67               45   
2             53278       61360  89.63         67               45   
3             57513       66113  88.48         67               45   
4             52378       60300  88.39         67               45   

  Employment.Type  State_ID  MobileNo_Avl_Flag  Aadhar_flag  PAN_flag   
0        Salaried         6                  1            1         0  \
1   Self employed         6                  1            1         0   
2   Self employed         6                  1            1         0   
3   Self employed         6                  1            1         0   
4   Self employed         6                  1            1         0   

   VoterID_flag  Driving_flag  Passport_flag  PERFORM_CNS.SCORE   
0             0             0              0                  0  \
1             0             0              0                598   
2             0             0              0                  0   
3             0             0              0                305   
4             0             0              0                  0   

  PERFORM_CNS.SCORE.DESCRIPTION  PRI.NO.OF.ACCTS  PRI.ACTIVE.ACCTS   
0   No Bureau History Available                0                 0  \
1                 I-Medium Risk                1                 1   
2   No Bureau History Available                0                 0   
3              L-Very High Risk                3                 0   
4   No Bureau History Available                0                 0   

   PRI.OVERDUE.ACCTS  PRI.CURRENT.BALANCE  PRI.SANCTIONED.AMOUNT   
0                  0                    0                      0  \
1                  1                27600                  50200   
2                  0                    0                      0   
3                  0                    0                      0   
4                  0                    0                      0   

   PRI.DISBURSED.AMOUNT  SEC.NO.OF.ACCTS  SEC.ACTIVE.ACCTS  SEC.OVERDUE.ACCTS   
0                     0                0                 0                  0  \
1                 50200                0                 0                  0   
2                     0                0                 0                  0   
3                     0                0                 0                  0   
4                     0                0                 0                  0   

   SEC.CURRENT.BALANCE  SEC.SANCTIONED.AMOUNT  SEC.DISBURSED.AMOUNT   
0                    0                      0                     0  \
1                    0                      0                     0   
2                    0                      0                     0   
3                    0                      0                     0   
4                    0                      0                     0   

   PRIMARY.INSTAL.AMT  SEC.INSTAL.AMT  NEW.ACCTS.IN.LAST.SIX.MONTHS   
0                   0               0                             0  \
1                1991               0                             0   
2                   0               0                             0   
3                  31               0                             0   
4                   0               0                             0   

   DELINQUENT.ACCTS.IN.LAST.SIX.MONTHS CREDIT.HISTORY.LENGTH  NO.OF_INQUIRIES   
0                                    0             0yrs 0mon                0  \
1                                    1            1yrs 11mon                0   
2                                    0             0yrs 0mon                0   
3                                    0             1yrs 3mon                1   
4                                    0             0yrs 0mon                1   

   AVERAGE.ACCT.AGE_year  AVERAGE.ACCT.AGE_month  
0                      0       

In [163]:
# extract year and month in separate columns from "AVERAGE.ACCT.AGE"
X_train = extract_month_year(X_train, "CREDIT.HISTORY.LENGTH")
X_test = extract_month_year(X_test, "CREDIT.HISTORY.LENGTH")
X_train.head()

disbursed_amount  asset_cost    ltv  branch_id  manufacturer_id   
0             50578       58400  89.55         67               45  \
1             47145       65550  73.23         67               45   
2             53278       61360  89.63         67               45   
3             57513       66113  88.48         67               45   
4             52378       60300  88.39         67               45   

  Employment.Type  State_ID  MobileNo_Avl_Flag  Aadhar_flag  PAN_flag   
0        Salaried         6                  1            1         0  \
1   Self employed         6                  1            1         0   
2   Self employed         6                  1            1         0   
3   Self employed         6                  1            1         0   
4   Self employed         6                  1            1         0   

   VoterID_flag  Driving_flag  Passport_flag  PERFORM_CNS.SCORE   
0             0             0              0                  0  \
1             0             0              0                598   
2             0             0              0                  0   
3             0             0              0                305   
4             0             0              0                  0   

  PERFORM_CNS.SCORE.DESCRIPTION  PRI.NO.OF.ACCTS  PRI.ACTIVE.ACCTS   
0   No Bureau History Available                0                 0  \
1                 I-Medium Risk                1                 1   
2   No Bureau History Available                0                 0   
3              L-Very High Risk                3                 0   
4   No Bureau History Available                0                 0   

   PRI.OVERDUE.ACCTS  PRI.CURRENT.BALANCE  PRI.SANCTIONED.AMOUNT   
0                  0                    0                      0  \
1                  1                27600                  50200   
2                  0                    0                      0   
3                  0                    0                      0   
4                  0                    0                      0   

   PRI.DISBURSED.AMOUNT  SEC.NO.OF.ACCTS  SEC.ACTIVE.ACCTS  SEC.OVERDUE.ACCTS   
0                     0                0                 0                  0  \
1                 50200                0                 0                  0   
2                     0                0                 0                  0   
3                     0                0                 0                  0   
4                     0                0                 0                  0   

   SEC.CURRENT.BALANCE  SEC.SANCTIONED.AMOUNT  SEC.DISBURSED.AMOUNT   
0                    0                      0                     0  \
1                    0                      0                     0   
2                    0                      0                     0   
3                    0                      0                     0   
4                    0                      0                     0   

   PRIMARY.INSTAL.AMT  SEC.INSTAL.AMT  NEW.ACCTS.IN.LAST.SIX.MONTHS   
0                   0               0                             0  \
1                1991               0                             0   
2                   0               0                             0   
3                  31               0                             0   
4                   0               0                             0   

   DELINQUENT.ACCTS.IN.LAST.SIX.MONTHS  NO.OF_INQUIRIES   
0                                    0                0  \
1                                    1                0   
2                                    0                0   
3                                    0                1   
4                                    0                1   

   AVERAGE.ACCT.AGE_year  AVERAGE.ACCT.AGE_month  CREDIT.HISTORY.LENGTH_year   
0                      0                       0                           0  \
1                      1                      11      

In [164]:
X_train = one_hot_encode(X_train, ["branch_id", "manufacturer_id", "Employment.Type", "State_ID", "PERFORM_CNS.SCORE.DESCRIPTION"])
X_test = one_hot_encode(X_test, ["branch_id", "manufacturer_id", "Employment.Type", "State_ID", "PERFORM_CNS.SCORE.DESCRIPTION"])

X_train.head()

disbursed_amount  asset_cost    ltv  MobileNo_Avl_Flag  Aadhar_flag   
0             50578       58400  89.55                  1            1  \
1             47145       65550  73.23                  1            1   
2             53278       61360  89.63                  1            1   
3             57513       66113  88.48                  1            1   
4             52378       60300  88.39                  1            1   

   PAN_flag  VoterID_flag  Driving_flag  Passport_flag  PERFORM_CNS.SCORE   
0         0             0             0              0                  0  \
1         0             0             0              0                598   
2         0             0             0              0                  0   
3         0             0             0              0                305   
4         0             0             0              0                  0   

   PRI.NO.OF.ACCTS  PRI.ACTIVE.ACCTS  PRI.OVERDUE.ACCTS  PRI.CURRENT.BALANCE   
0                0                 0                  0                    0  \
1                1                 1                  1                27600   
2                0                 0                  0                    0   
3                3                 0                  0                    0   
4                0                 0                  0                    0   

   PRI.SANCTIONED.AMOUNT  PRI.DISBURSED.AMOUNT  SEC.NO.OF.ACCTS   
0                      0                     0                0  \
1                  50200                 50200                0   
2                      0                     0                0   
3                      0                     0                0   
4                      0                     0                0   

   SEC.ACTIVE.ACCTS  SEC.OVERDUE.ACCTS  SEC.CURRENT.BALANCE   
0                 0                  0                    0  \
1                 0                  0                    0   
2                 0                  0                    0   
3                 0                  0                    0   
4                 0                  0                    0   

   SEC.SANCTIONED.AMOUNT  SEC.DISBURSED.AMOUNT  PRIMARY.INSTAL.AMT   
0                      0                     0                   0  \
1                      0                     0                1991   
2                      0                     0                   0   
3                      0                     0                  31   
4                      0                     0                   0   

   SEC.INSTAL.AMT  NEW.ACCTS.IN.LAST.SIX.MONTHS   
0               0                             0  \
1               0                             0   
2               0                             0   
3               0                             0   
4               0                             0   

   DELINQUENT.ACCTS.IN.LAST.SIX.MONTHS  NO.OF_INQUIRIES   
0                                    0                0  \
1                                    1                0   
2                                    0                0   
3                                    0                1   
4                                    0                1   

   AVERAGE.ACCT.AGE_year  AVERAGE.ACCT.AGE_month  CREDIT.HISTORY.LENGTH_year   
0                      0                       0                           0  \
1                      1                      11                           1   
2                      0                       0                           0   
3                      0                       8                           1   
4                      0                       0                           0   

   CREDIT.HISTORY.LENGTH_month  branch_id_1  branch_id_2  branch_id_3   
0                            0            0            0            0  \
1                           11            0            0            0   
2                            0            0      

Checking for non-identical columns in train set and test set.

In [168]:
# columns which are present in train set but not in test set

extra_train_set_col = []
extra_test_set_col = []

print("--columns in train set but not in test set--")
for column in X_train.columns:
    if column not in X_test.columns:
        print(column)
        extra_train_set_col.append(column)
print("\n")


print("--columns in test set but not in train set--")
for column in X_test.columns:
    if column not in X_train.columns:
        print(column)
        extra_test_set_col.append(column)


--columns in train set but not in test set--
manufacturer_id_156
PERFORM_CNS.SCORE.DESCRIPTION_Not Scored: More than 50 active Accounts found


--columns in test set but not in train set--
manufacturer_id_155


Removing columns that are present in test set but not in train set because model will be trained in train set.

In [169]:
X_test = X_test.drop(extra_test_set_col, axis=1)

Adding columns in test set which are not present in test set.

In [170]:
for column in extra_train_set_col:
    X_test[column] = 0
    
